### DISCUSSION TITLE
[HW2] problem 2.3

### DISCUSSION THREAD

**Discussion by KimSSung** [seominjoon/kaist-ai605#128] [Q&A] [answered]
Hello!
My search method takes too much time (30 minutes + @, so I canceled)... can I get some tips?

- I'm using cpu (since I'm not sure where to use .to(device))
- I'm using 'context' of SQUAD train dataset for **train** , valid dataset for **add** and **search**
- **search** calculates sum of **tf-idf** of each term in query (also calculate **tf** of each term in here)

I'm so worried that I think I'm doing something wrong..
Thanks.

**Comment by sorobedio** [comment, marked as answer]
i did same as your first two steps. i don't know what you mean by sum of tf-idf but you need to compute tf for every term in a query and that is not computationally expensive than the tf-idf computation during training process. In my case tf-idf computation is less than 1 or 2 minutes.  you may want to check which part of your code in the search function is exactly slow.

**Reply by KimSSung** [reply]
oh.... it seems that my understanding of tf-idf is wrong.
![image](https://user-images.githubusercontent.com/39760214/138803962-9877943a-f931-43a5-9a83-55923660ff6f.png)
I thought that expression below is what I have to use in tf-idf search method (for each document, get SUM of tf(term, document) * idf(term) for all terms in the query -> then find documents of largest 10 scores).
May be this is wrong right?

**Reply by sorobedio** [reply]
The top formula you posted is BM25 not tf-idf. 

This is tf-idf

![tf-idf](https://user-images.githubusercontent.com/70527807/138809334-994d744b-68aa-4bd1-b20d-091e011e6620.PNG)


which one are you talking about

**Reply by KimSSung** [reply]
Yes, I used it, but then how to find what document is most similar with query?
Is this right ?
- [Bag of words]  cossine similarity(BoW vector of query, BoW vector of document) -> find top k document of this cos-sim
- [TF-IDF] score: sum TF(term, document) * IDF(term) for all terms in query -> find top k document of this score

(My understanding is so slow since it's quite first time to learn nlp,,,, Thank you for helping me!)

**Reply by sorobedio** [reply]
if you understand BOW similarity search you understand tf-idf similarity search because it is the same BOW vectors (same dimension) only the value of each term is changed to its tf-idf value.  That is how i understood.

**Reply by KimSSung** [reply]
ah-ha you means that when
vocab = ['a', 'b', 'c', 'd']
document1 = 'a b c'
document2 = 'b c d'
query = 'a c d'

then

- BoW: [1,1,1,0] for doc1, [0,1,1,1] for doc2, [1,0,1,1] for query -> cossim(doc1, query) & cossim(doc2, query) find bigger doc
- TF-IDF: [TF('a', doc1)* IDF('a'), TF('b', doc1)* IDF('b'), TF('c', doc1)* IDF('c'), TF('d', doc1)* IDF('d')], and same for doc2 -> then cossim(doc1, query) ......

am I right? :)

**Reply by sorobedio** [reply]
yes something like that.

**Reply by KimSSung** [reply]
I'll try it, thank you very much !
And sorry for the additional question, can I ask how much time it takes for 10570 valid documents (Roughly) ?

**Reply by sorobedio** [reply]
less than 5 minute for sure
